A39-xds-http-filters.md

```diff
@@ -0,0 +1,145 @@
+response, it will validate the list of filters as follows:
+- Every filter in the list must be of a known type, as identified by the
+  `type_url` field.  Any unknown config type will cause the Listener
+  resource to be NACKed.
+- Every filter in the list must have a unique instance name.  Any
+  duplicate instance names will cause the Listener resource to be NACKed.
+
+Note that gRPC will support configuring multiple instances of the same
+filter implementation, each with its own name and config.
+
+In each of the `VirtualHost`, `Route`, and `WeightedCluster.ClusterWeight`
+protos, there is a field called `typed_per_filter_config`, which is a map
+from filter instance name to a corresponding filter-specific configuration
+proto message.  (Note that Envoy currently keys these maps by legacy
+filter names instead of by filter instance names, but this is expected
+to change, as per https://github.com/envoyproxy/envoy/issues/12274; gRPC
+will directly implement the desired semantics here.)
+
+When gRPC receives the `RouteConfiguration` proto, either in the LDS or
+RDS response, it will validate the contents of each
+`typed_per_filter_config` map as follows:
+- Every config in the list must be of a known type, as identified by the
+  map value proto message type.  Any unknown config type will cause the
+  resource to be NACKed.
+- Note that gRPC will *not* fail validation if the map key specifies a
+  filter instance name that does not exist in the `HttpConnectionManager`
+  filter list.  This is because during an update, the xDS client code
+  cannot know which `HttpConnectionManager` config is currently being
+  used.
+
+### Initially Supported Filter: Router
+
+Initially, gRPC will support only one filter, the [router
+filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter),
+which is used for the proto message type
+[`extensions.filters.http.router.v3.Router`](https://github.com/envoyproxy/envoy/blob/18db4c90e3295fb2c39bfc7b2ce641cfd6c3fbed/api/envoy/extensions/filters/http/router/v3/router.proto#L23).
+As in Envoy, this filter is required and must be last in the filter list.
+
+Note that gRPC will not actually have a discrete filter implementation
+for the router filter; instead, this filter will simply trigger the
+built-in routing functionality that we have already implemented in the
+resolver and LB policy plugins to the gRPC client channel.  However,
+in order to retain compatibility with Envoy, gRPC will still accept
+this filter configuration and treat it similarly to the way that Envoy
+does.
+
+Specifically, if the router filter is not present, gRPC will fail all
+requests.  In addition, any filters in the list after the router filter
+will not actually be used, although their configurations will still be
+validated when the config is received from the xDS server.
+
+For now, gRPC will not support any of the fields in the router filter's
+config.  All fields will be ignored.
```

Comment:
Should we just ignore the fields or NACK the config if it contains values we don't support? e.g. if `dynamic_stats` is true?

Reply:
We ignore them.  This is no different than any other individual xDS field we don't (or don't *yet*) support.